include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -529,385 +477,362 @@ vec3 fetchTex(in uvec3 texid, in vec2 uv, in mat2 dUV)
-	currInteraction.B = -currInteraction.B;
-	currInteraction.TdotV = -currInteraction.TdotV;
-	currInteraction.BdotV = -currInteraction.BdotV;
-}
-//call before executing an instruction/evaluating bsdf
-void handleTwosided_interactionOnly(inout bool ts_flag, in instr_t instr)
+#ifdef GEN_CHOICE_STREAM
+void instr_eval_and_pdf_execute(in instr_t instr, in MC_precomputed_t precomp, in irr_glsl_LightSample s, in MC_microfacet_t _microfacet, in bool skip)
 {
-#ifndef NO_TWOSIDED
-	if (!ts_flag && instr_getTwosided(instr))
+	const uint op = instr_getOpcode(instr);
+	const bool is_bxdf = op_isBXDF(op);
+	const bool is_bsdf = !op_isBRDF(op); //note it actually tells if op is BSDF or BUMPMAP or SET_GEOM_NORMAL (divergence reasons) [(is_bxdf && is_bsdf) actually tells that op is bsdf]
+	// (is_bxdf && is_bsdf) -> BSDF
+	// (is_bxdf && !is_bsdf) -> BRDF
+	const bool is_bxdf_or_combiner = op_isBXDForCoatOrBlend(op);
+
+	uvec3 regs = instr_decodeRegisters(instr);
+	params_t params;
+	bsdf_data_t bsdf_data;
+	mat2x3 ior;
+	mat2x3 ior2;
+	MC_microfacet_t microfacet;
+
+	const bool run = !skip;
+
+	if (is_bxdf_or_combiner && run)
 	{
-		ts_flag = true;
-		if (currInteraction.isotropic.NdotV<0.0)
-		{
-			flipCurrInteraction();
-		}
+		bsdf_data = fetchBSDFDataForInstr(instr);
+		ior = bsdf_data_decodeIoR(bsdf_data, op);
+		ior2 = matrixCompMult(ior, ior);
+		params = instr_getParameters(instr, bsdf_data);
 	}
-#ifdef OP_BUMPMAP
-	ts_flag = instr_getOpcode(instr)==OP_BUMPMAP ? false:ts_flag;
-#endif
-#endif	
-}
-//call before executing an instruction/evaluating bsdf
-void handleTwosided(inout bool ts_flag, in instr_t instr)
-{
-#ifndef NO_TWOSIDED
-	if (!ts_flag && instr_getTwosided(instr))
+
+	const float NdotV = irr_glsl_conditionalAbsOrMax(is_bsdf, currInteraction.inner.isotropic.NdotV, 0.0);
+
+	vec3 eval = vec3(0.0);
+	float pdf = 0.0;
+
+	if (is_bxdf && run && (NdotV > FLT_MIN))
 	{
-		ts_flag = true;
-		if (currInteraction.isotropic.NdotV<0.0)
+		//speculative execution
+		uint ndf = instr_getNDF(instr);
+		float a = params_getAlpha(params);
+		float a2 = a*a;
+#ifdef ALL_ISOTROPIC_BXDFS
+		float one_minus_a2 = 1.0 - a2;
+#else
+		float ay = params_getAlphaV(params);
+		float ay2 = ay*ay;
+#endif
+
+		const vec3 albedo = params_getReflectance(params);
+
+		const float NdotL = irr_glsl_conditionalAbsOrMax(is_bsdf, s.NdotL, 0.0);
+
+#if defined(OP_DIFFUSE) || defined(OP_DIFFTRANS)
+		if (op_isDiffuse(op))
 		{
-			flipCurrInteraction();
-
-			currBSDFParams.isotropic.NdotL = -currBSDFParams.isotropic.NdotL;
-			currBSDFParams.isotropic.NdotH = -currBSDFParams.isotropic.NdotH;
-			currBSDFParams.TdotL = -currBSDFParams.TdotL;
-			currBSDFParams.BdotL = -currBSDFParams.BdotL;
-			currBSDFParams.TdotH = -currBSDFParams.TdotH;
-			currBSDFParams.BdotH = -currBSDFParams.BdotH;
+			eval = albedo * irr_glsl_oren_nayar_cos_remainder_and_pdf_wo_clamps(pdf, a2, s.VdotL, NdotL, NdotV);
+			pdf *= is_bsdf ? 0.5 : 1.0;
+			eval *= pdf;
```

Comment:
put another if-statement inside to check for `NdotL >FLT_MIN`

https://github.com/buildaworldnet/IrrlichtBAW/pull/554#discussion_r528997756